// tfjs-onnx/dist/node/layers/activations.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var tf = require("@tensorflow/tfjs");
var node_1 = require("../node");
var Activation = (function (_super) {
    __extends(Activation, _super);
    function Activation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Activation.prototype.getTfjsLayer = function (node) {
        return tf.layers.activation(this.getTfjsConfig(node));
    };
    return Activation;
}(node_1.OnnxNode));
exports.Activation = Activation;
var Relu = (function (_super) {
    __extends(Relu, _super);
    function Relu() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Relu.prototype.getTfjsLayerConfig = function (node) {
        return { activation: 'relu' };
    };
    return Relu;
}(Activation));
exports.Relu = Relu;
var Tanh = (function (_super) {
    __extends(Tanh, _super);
    function Tanh() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Tanh.prototype.getTfjsLayerConfig = function (node) {
        return { activation: 'tanh' };
    };
    return Tanh;
}(Activation));
exports.Tanh = Tanh;
var Sigmoid = (function (_super) {
    __extends(Sigmoid, _super);
    function Sigmoid() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Sigmoid.prototype.getTfjsLayerConfig = function (node) {
        return { activation: 'sigmoid' };
    };
    return Sigmoid;
}(Activation));
exports.Sigmoid = Sigmoid;
var Elu = (function (_super) {
    __extends(Elu, _super);
    function Elu() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Elu.prototype.getTfjsLayerConfig = function (node) {
        return { activation: 'elu' };
    };
    return Elu;
}(Activation));
exports.Elu = Elu;
var Softplus = (function (_super) {
    __extends(Softplus, _super);
    function Softplus() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Softplus.prototype.getTfjsLayerConfig = function (node) {
        return { activation: 'softplus' };
    };
    return Softplus;
}(Activation));
exports.Softplus = Softplus;
var Softsign = (function (_super) {
    __extends(Softsign, _super);
    function Softsign() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Softsign.prototype.getTfjsLayerConfig = function (node) {
        return { activation: 'softsign' };
    };
    return Softsign;
}(Activation));
exports.Softsign = Softsign;
var HardSigmoid = (function (_super) {
    __extends(HardSigmoid, _super);
    function HardSigmoid() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HardSigmoid.prototype.getTfjsLayerConfig = function (node) {
        return { activation: 'hardsigmoid' };
    };
    return HardSigmoid;
}(Activation));
exports.HardSigmoid = HardSigmoid;
