// tfjs-onnx/dist/node/node.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var initializers_1 = require("@tensorflow/tfjs-layers/dist/initializers");
var util = require("./util");
var OnnxNode = (function () {
    function OnnxNode(model) {
        this.model = model;
    }
    ;
    OnnxNode.from = function (model) {
        var that = new this(model);
        return that;
    };
    OnnxNode.prototype.getTfjsConfig = function (node, input) {
        var commonConfig = util.getTfjsCommonConfig(node);
        var layerConfig = this.getTfjsLayerConfig(node, input);
        return Object.assign({}, commonConfig, layerConfig);
    };
    return OnnxNode;
}());
exports.OnnxNode = OnnxNode;
var WeightInitializer = (function (_super) {
    __extends(WeightInitializer, _super);
    function WeightInitializer(weights) {
        var _this = _super.call(this) || this;
        _this.weights = weights;
        return _this;
    }
    ;
    WeightInitializer.prototype.apply = function (shape, dtype) {
        return this.weights;
    };
    WeightInitializer.prototype.getClassName = function () {
        return 'WeightInitializer';
    };
    return WeightInitializer;
}(initializers_1.Initializer));
exports.WeightInitializer = WeightInitializer;
